Add LocalFileConfigLoader

thomashorrobin opened the pull request:
Hi, I need a function like this so that my application can load user selected files

- Why not just build the extension in your own app? `AuthInfo.authentication` and `NamedCluster.trustRoots` are both private functions

- Why would anyone need this, it seems like an objectively worse way of loading a client. Issues to do with the Sandbox, mainly. But also, it is useful to be able to open a config directly.

I don't have a lot of experience with Swift, so I may have done this all wrong. Feel free to set me straight if my approach is wrong.

thomashorrobin committed:
added loadConfigFromPath

hung loadConfigFromPath off of KubernetesClientConfig

iabudiab commented on Sources/SwiftkubeClient/Config/KubernetesClientConfig.swift:
```diff
@@ -228,3 +228,51 @@ private extension AuthInfo {
 }
+
+public extension KubernetesClientConfig {
+	static func loadConfigFromPath(logger: Logger, path: String) throws -> KubernetesClientConfig? {
```
You could change the function to look more "swifty" by moving the `path` parameter to first position and dropping the `fromPath` part:

```swift
func load(fromPath: String, logger: Logger) throws -> KubernetesClientConfig? { ... }
``` 

The call-site would look like this then:

```swift
KubernetesClientConfig().load(fromPath: "/some/path/", logger: logger)
```

thomashorrobin replied:
ah yes, that reads much better! I will implement 👍

iabudiab commented on Sources/SwiftkubeClient/Config/KubernetesClientConfig.swift:
```diff
@@ -228,3 +228,51 @@ private extension AuthInfo {
+		guard let contents = try? String(contentsOf: fileURL, encoding: .utf8) else {
+			return nil
+		}
+
+		guard let kubeConfig = try? decoder.decode(KubeConfig.self, from: contents) else {
+			return nil
+		}
+
+		guard let currentContext = kubeConfig.currentContext else {
+			return nil
+		}
+
+		guard let context = kubeConfig.contexts?.filter({ $0.name == currentContext }).map(\.context).first else {
+			return nil
+		}
+
+		guard let cluster = kubeConfig.clusters?.filter({ $0.name == context.cluster }).map(\.cluster).first else {
+			return nil
+		}
+
+		guard let masterURL = URL(string: cluster.server) else {
+			return nil
+		}
+
+		guard let authInfo = kubeConfig.users?.filter({ $0.name == context.user }).map(\.authInfo).first else {
+			return nil
+		}
+
+		guard let authentication = authInfo.authentication(logger: logger) else {
+			return nil
+		}
+
+		return KubernetesClientConfig(
+			masterURL: masterURL,
+			namespace: context.namespace ?? "default",
+			authentication: authentication,
+			trustRoots: cluster.trustRoots(logger: logger),
+			insecureSkipTLSVerify: cluster.insecureSkipTLSVerify ?? true
+		)
+	}
```
This is an exact duplicate of the same lines in the `LocalFileConfigLoader`. How about this:

- We change the `LocalFileConfigLoader` to accept a `URL` and the body would contains this logic for loading any given arbitrary `URL`.

```swift
internal struct LocalFileConfigLoadera {
   internal func load(fromURL: URL, logger: Logger) throws -> KubernetesClientConfig? { ... }
} 
```
- Then we rename `LocalFileConfigLoader` to something like `URLConfigLoader`
- And finally rename the new `KubernetesClientConfig` to something like `LocalKubeConfigLoader` and make it call the `URLConfigLoader` to restore the previous functionality, e.g.:

```swift
struct LocalKubeConfigLoader {
    func load(logger: Logger) throws -> KubernetesClientConfig? {
        guard let homePath = ProcessInfo.processInfo.environment["HOME"] else {
            logger.info("Skipping kubeconfig in $HOME/.kube/config because HOME env variable is not set.")
            return nil
        }
        let kubeConfigURL = URL(fileURLWithPath: homePath + "/.kube/config")
        return try? URLConfigLoader().load(fromURL: kubeConfigURL, logger: logger)
    }
}
```

thomashorrobin replied:
Ah I see. So we will have three structs the implement `KubernetesClientConfigLoader`
1. `LocalFileConfigLoader`
2. `LocalKubeConfigLoader`
3. `ServiceAccountConfigLoader`
Cool, I reckon that'll work well 😊

iabudiab replied:
Yes exactly, however the `LocalFileConfigLoader` can accept any `URL`, even remote URLs` (don't know if that is really a use-case), so we could just call it `URLConfigLoader` 😉

iabudiab commented on Sources/SwiftkubeClient/Config/KubernetesClientConfig.swift:
```diff
@@ -228,3 +228,51 @@ private extension AuthInfo {
 	}
 }
+
+public extension KubernetesClientConfig {
```
Can you make this implement the `KubernetesClientConfigLoader` protocol analogous to the other loaders. It would however mean that we would have to create an instance of the struct and, as of now, it is a one-shot-then-throw-away instance (as you can see [here](https://github.com/swiftkube/client/blob/f40782168edb0f466448b744df4a51333a4c05d1/Sources/SwiftkubeClient/Client/KubernetesClient.swift#L103)).

thomashorrobin replied:
Ah yes, I did actually consider this when I first looked at it, but the one-shot-then-throw-away thing confused me. But yeah, if nothing else it keeps everything consistent and maintainable. I will change it

iabudiab reviewed (commented):
Take a look at the last comment first since it has the most impact 😉

iabudiab commented:
@thomashorrobin Hey there 👋 welcome to the project 😁 and thanks for the PR, looks good!

I left a couple of comments there with some points, nothing's wrong just some potential for improvement 😉 

Cheers

thomashorrobin committed:
extract logic to load URLs into URLConfigLoader and added LocalKubeConfigLoader plus new convenience constructor

This commit removes the KubernetesClientConfig.loadURL extension and moves the functionality to a struct that implements KubernetesClientConfigLoader. This is so the loaders are consistent which will help future proof the project. Also a convenience constructor has been added to take advantage of the new functionality in URLConfigLoader

thomashorrobin commented:
Hi @iabudiab, I accidentally requested a review for this PR and now I seems that I can't undo that. However it's not quite ready for review, I will let you know I the comments when it is 😊

thomashorrobin commented:
Hi @iabudiab, I've implemented everything in review with two exceptions. I decided to keep `LocalFileConfigLoader` as an internal struct and expose the functionality through a new constructor on `KubernetesClient`. I personally think this is tidier than making it public and then having developers putting back in using `KubernetesClient(config: config)`

Also, in the new constructor I've put in a guard statement to throw if the URL doesn't point to a file. This is contrary to what we discussed but I when I was testing it on a config file from the web I ran into all sorts of networking issues. This thread best describes the issue: https://github.com/airbnb/lottie-ios/issues/1256 I tried a number of things to fix this but nothing really worked.

thomashorrobin committed:
fixed issues whereby non file URLs were silently breaking

thomashorrobin commented:
@iabudiab I've fixed the issues with networking using a non-file URL. Let me know if you want more details

thomashorrobin committed:
ran swiftformat

iabudiab commented:
@thomashorrobin really sorry, I got distracted yesterday 😅 and didn't finish the PR. Looks good now 👍
